libraries/AP_Generator/AP_RichenPower.cpp

```diff
@@ -0,0 +1,430 @@
+    last_reading.output_current = be16toh(u.packet.output_current) / 100.0f;
+    last_reading.mode = (Mode)u.packet.mode;
+
+    last_reading_ms = AP_HAL::millis();
+
+    body_length = 0;
+
+    return true;
+}
+
+void AP_RichenPower::update_heat()
+{
+    // assume heat increase is directly proportional to RPM.
+    const uint32_t now = AP_HAL::millis();
+    uint16_t rpm = last_reading.rpm;
+    if (now - last_reading_ms > 2000) {
+        // if we're not getting updates, assume we're getting colder
+        rpm = 0;
+        // ... and resend the version information when we get something again
+        protocol_information_anounced = false;
+    }
+
+    const uint32_t time_delta_ms = now - last_heat_update_ms;
+    last_heat_update_ms = now;
+
+    heat += rpm * time_delta_ms * (1/1000.0f);
+    // cap the heat of the motor:
+    heat = MIN(heat, 60 * RUN_RPM); // so cap heat at 60 seconds at run-speed
+    // now lose some heat to the environment
+    heat -= (heat * environment_loss_factor * (time_delta_ms * (1/1000.0f)));  // lose some % of heat per second
+
+    // if (now - last_heat_print > 100) {
+    //     last_heat_print = now;
+    //     ::fprintf(stderr, "Heat: %f    rpm:%u\n", heat, rpm);
+    // }
+}
+
+constexpr float AP_RichenPower::heat_required_for_run()
+{
+    // assume that heat is proportional to RPM.  Return a number
+    // proportial to RPM.  Reduce it to account for the cooling some%/s
+    // cooling
+    return (30 * IDLE_RPM) * powf(1.0f-environment_loss_factor, 30);
+}
+bool AP_RichenPower::generator_ok_to_run() const
+{
+    return heat > heat_required_for_run();
+}
+
+constexpr float AP_RichenPower::heat_required_for_supply()
+{
+    // account for cooling that happens in that 60 seconds
+    return (30 * IDLE_RPM + 30 * RUN_RPM) * powf((1.0f-environment_loss_factor), 60);
+}
+bool AP_RichenPower::generator_ok_to_supply() const
+{
+    // duplicated into prearms
+    return heat > heat_required_for_supply();
+}
+
+
+
+/*
+  update the state of the sensor
+*/
+void AP_RichenPower::update(void)
+{
+    if (uart == nullptr) {
+        return;
+    }
+
+    update_servo_channel();
+
+    update_runstate();
+
+    if (get_reading()) {
+        // update
+        //printf("Data is %.2f\r\n",hybridData);
+    } else if (AP_HAL::millis() - last_reading_ms > 2000) {
+//        printf("No hybrid data\r\n");
+    }
+
+    update_heat();
+
+    Log_Write();
+}
+
+void AP_RichenPower::update_runstate()
+{
+    if (_servo_channel == nullptr) {
+        // doesn't really matter what we want to do; we're not going
+        // to be able to affect it...
+        return;
+    }
+
+    // don't run the generator while the safety is on:
+    // if (hal.util->safety_switch_state() == AP_HAL::Util::SAFETY_DISARMED) {
+    //     _servo_channel->set_output_pwm(SERVO_PWM_STOP);
+    //     return;
+    // }
+
+    static const uint16_t SERVO_PWM_STOP = 1200;
```

Comment:
I think we normally put these kinds of static definition-ish things in the .h or top of the .cpp don't we?